// @project-sunbird/sb-dashlet/esm2020/lib/dashlet.module.mjs
import { NgModule } from '@angular/core';
import { DashletComponent, ChartJsComponent, BigNumberComponent, DtTableComponent, FiltersComponent, MapComponent } from './components/index';
import { ReportWrapperDirective, TemplateRefsDirective } from './directives/index';
import { HttpClientModule } from '@angular/common/http';
import { ChartsModule } from 'ng2-charts';
import { CommonModule } from '@angular/common';
import { DataTablesModule } from "angular-datatables";
import { ReactiveFormsModule } from '@angular/forms';
import { NgMultiSelectDropDownModule } from 'ng-multiselect-dropdown';
import { NgxDaterangepickerMd } from 'ngx-daterangepicker-material';
import { DATA_SERVICE } from './tokens/index';
import { DataService } from './services/index';
import * as i0 from "@angular/core";
import * as i1 from "ngx-daterangepicker-material";
export class DashletModule {
    static forRoot(config) {
        return {
            ngModule: DashletModule,
            providers: [
                { provide: DATA_SERVICE, useClass: (config && config.dataService) || DataService }
            ]
        };
    }
}
DashletModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: DashletModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DashletModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: DashletModule, declarations: [ChartJsComponent, DashletComponent, ReportWrapperDirective, BigNumberComponent, DtTableComponent, TemplateRefsDirective, FiltersComponent, MapComponent], imports: [HttpClientModule, ChartsModule, CommonModule, DataTablesModule, ReactiveFormsModule, NgMultiSelectDropDownModule, i1.NgxDaterangepickerMd], exports: [DashletComponent, TemplateRefsDirective] });
DashletModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: DashletModule, imports: [HttpClientModule, ChartsModule, CommonModule, DataTablesModule, ReactiveFormsModule, NgMultiSelectDropDownModule,
        NgxDaterangepickerMd.forRoot()] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: DashletModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ChartJsComponent, DashletComponent, ReportWrapperDirective, BigNumberComponent, DtTableComponent, TemplateRefsDirective, FiltersComponent, MapComponent],
                    imports: [HttpClientModule, ChartsModule, CommonModule, DataTablesModule, ReactiveFormsModule, NgMultiSelectDropDownModule,
                        NgxDaterangepickerMd.forRoot(),
                    ],
                    exports: [DashletComponent, TemplateRefsDirective]
                }]
        }] });
