// cqube-library/esm2020/lib/cqube-library.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CqubeLibraryComponent } from './cqube-library.component';
import { CqubeInfoCardComponent } from './cqube-info-card/cqube-info-card.component';
import { CqubeProgramCardComponent } from './cqube-program-card/cqube-program-card.component';
import * as i0 from "@angular/core";
export class CqubeLibraryModule {
}
CqubeLibraryModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.5", ngImport: i0, type: CqubeLibraryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
CqubeLibraryModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.0.5", ngImport: i0, type: CqubeLibraryModule, declarations: [CqubeLibraryComponent,
        CqubeInfoCardComponent,
        CqubeProgramCardComponent], imports: [CommonModule], exports: [CqubeLibraryComponent,
        CqubeInfoCardComponent,
        CqubeProgramCardComponent] });
CqubeLibraryModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.0.5", ngImport: i0, type: CqubeLibraryModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.5", ngImport: i0, type: CqubeLibraryModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        CqubeLibraryComponent,
                        CqubeInfoCardComponent,
                        CqubeProgramCardComponent
                    ],
                    imports: [
                        CommonModule
                    ],
                    exports: [
                        CqubeLibraryComponent,
                        CqubeInfoCardComponent,
                        CqubeProgramCardComponent
                    ]
                }]
        }] });
